lib/mercury/received_message.rb

```diff
@@ -18,17 +18,31 @@ def headers
 
     def ack
       @is_ackable or raise 'This message is not ackable'
+      disallow_double_actions
+      @action_taken = :ack
       metadata.ack
     end
 
     def reject
       @is_ackable or raise 'This message is not rejectable'
+      disallow_double_actions
+      @action_taken = :reject
       metadata.reject(requeue: false)
     end
 
     def nack
       @is_ackable or raise 'This message is not nackable'
+      disallow_double_actions
+      @action_taken = :nack
       metadata.reject(requeue: true)
     end
+
+    private
+
+    def disallow_double_actions
+      if @action_taken
+        raise "This message was already #{@action_taken}ed"
+      end
+    end
   end
 end
```

Comment:
You could force the accessing / setting of `@action_taken` to occur in the `disallow_double_actions` by doing something like this

```ruby
def set_action(action)
  if @action_taken
    raise "This message was already #{@action_taken}ed"
  end
  @action_taken = action_to_set
end
```

Then you can get rid of the `@action_taken =` from each of your public methods.